include/dmlc/logging.h

```diff
@@ -162,51 +162,19 @@ inline bool DebugLoggingEnabled() {
   return state == 1;
 }
 
-class LogCheckError {
- public:
-  LogCheckError() : str(nullptr) {}
-  explicit LogCheckError(const std::string& str_) : str(new std::string(str_)) {}
-  LogCheckError(const LogCheckError& other) = delete;
-  LogCheckError(LogCheckError&& other) : str(other.str) {
-    other.str = nullptr;
-  }
-  ~LogCheckError() { if (str != nullptr) delete str; }
-  operator bool() const { return str != nullptr; }
-  LogCheckError& operator=(const LogCheckError& other) = delete;
-  LogCheckError& operator=(LogCheckError&& other) = delete;
-  std::string* str;
-};
-
 #ifndef DMLC_GLOG_DEFINED
 
-#ifndef _LIBCPP_SGX_NO_IOSTREAMS
-#define DEFINE_CHECK_FUNC(name, op)                               \
-  template <typename X, typename Y>                               \
-  inline LogCheckError LogCheck##name(const X& x, const Y& y) {   \
-    if (x op y) return LogCheckError();                           \
-    std::ostringstream os;                                        \
-    os << " (" << x << " vs. " << y << ") ";  /* CHECK_XX(x, y) requires x and y can be serialized to string. Use CHECK(x OP y) otherwise. NOLINT(*) */ \
-    return LogCheckError(os.str());                               \
-  }                                                               \
-  inline LogCheckError LogCheck##name(int x, int y) {             \
-    return LogCheck##name<int, int>(x, y);                        \
-  }
-#else
+// This function allows us to ignore sign comparison in the right scope.
+// Once dmlc requires c++11, #pragma can be changed to _Pramga, and directives
+// can be moved to CHECK_BINARY_OP definition itself
 #define DEFINE_CHECK_FUNC(name, op)                               \
   template <typename X, typename Y>                               \
-  inline LogCheckError LogCheck##name(const X& x, const Y& y) {   \
-    if (x op y) return LogCheckError();                           \
-    return LogCheckError("Error.");                               \
+  inline bool LogCheck##name(const X& x, const Y& y) {            \
+    return (x op y);                                              \
```

Comment:
shall we always inline this check? given that this is now small

Reply:
Sure. Is it correct that dmlc-core needs to support pre cpp11? In cpp11 it should be easy to put the condition into the macro itself, and guard just that condition with ignored "-Wsign-compare".

Reply:
We now requires cpp11. However, we do need to make sure that the code is compiler agnostic.

Reply:
putting it directly to macro appears trickier than i thought, because 
1) macro doesn't expand to complete statement, so that one can write CHECK_GE(a, b) << "bad thing";
2) we probably want to keep ```GCC diagnostic ignored "-Wsign-compare"``` around condition itself only. 

One option could be to create a temporary variable, so the code in CHECK_BINARY_OP  changes to something like this:

```diff
-if (!(dmlc::LogCheck##name(x, y))) 
+_Pragma("GCC diagnostic push");
+_Pragma("GCC diagnostic ignored "-Wsign-compare"");
+ auto some_unique_name##__LINE__ = !((x) op (y));
+_Pragma("GCC diagnostic pop");
+if (some_unique_name##__LINE__) 
```

This doesn't look nice, + name collisions are still possible; ```__COUNTER__``` is not part of the standard. 

Another option is to define always_inline attribute, similar to how it's done for noinline here: https://github.com/dmlc/dmlc-core/blob/dc0156a73a846751851fcf130729ade6360a37b1/include/dmlc/base.h#L239-L243

and mark the function with that attribute.

Which would you prefer?
Any other options I'm missing?

Reply:
Let us go with always inline then